File: leetcodes/226. Invert Binary Tree/invertBinaryTree.java
import java.util.Queue;

import javax.swing.tree.TreeNode;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  public TreeNode invertTree(TreeNode root) {
    //bfs time O(h) -> worse case O(n), space O(n)
    //if root is null
    if (root == null) return root;
    //create queue
    final Queue<TreeNode> queue = new LinkedList<>();
    //offer root into queue
    queue.offer(root);
    //while queue is not empty
    while (!queue.isEmpty()){
      //poll first elem from queue
      final TreeNode node = queue.poll();
      final TreeNode left = node.left;
      //assign its left to right
      //assign its right to left
      node.left = node.right;
      node.right = left;
      //if its left is not null -> offer to queue
      if (node.left != null) queue.offer(node.left);
      //if its right is not null offer to queue
      if (node.right != null) queue.offer(node.right);
    }
    return root;
  }
}

File: leetcodes/226. Invert Binary Tree/invertBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
      #return 0
      '''
       BFS
          -> exchange left and right for each node and return the root
          space O(1), time O(h) -> worse case O(n)-> one sided tree
       DFS time O(h), spaceO(h)
          -> if !root: return null
          left = root.right
          root.left = dfs(root.right)
          root.right = dfs(left)
          return root
      '''
      if root == None:
        return root
      queue = [root]
      while len(queue):
        node = queue.pop()
        #swapping first -> append swapped nodes later
        Left = node.left
        Right = node.right
        node.right = Left
        node.left = Right
        #don't append None node into the queue
        if node.right != None:
            queue.append(Left)
        if node.left != None:
            queue.append(Right)
      return root
#DFS recursively
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
      if root == None:
        return None
      Left = self.invertTree(root.right)
      Right = self.invertTree(root.left)
      root.left = Left
      root.right = Right
      return root